src/stack.ts

```diff
@@ -11,172 +11,252 @@
+                (prev, curr) => {
+                    const o: { [outputId: string]: pulumi.Output<any> } = {};
+                    for (const [outputId, args] of Object.entries(curr.stack.outputs ?? {})) {
+                        o[outputId] = curr.processIntrinsics(args.Value);
+                    }
+                    return {
+                        ...prev,
+                        ...o,
+                    };
+                },
+                { ...this.outputs } as pulumi.Output<{ [outputId: string]: pulumi.Output<any> }>,
+            );
+        });
+        this.outputs = pulumi.output(outputs);
+        this.registerOutputs(this.outputs);
+    }
 
     /**
      * @internal
      */
-    public dependencies: CdkConstruct[] = [];
+    public get app(): cdk.App {
+        if (!this._app) {
+            throw new Error('cdk.App has not been created yet');
+        }
+        return this._app!;
+    }
 
-    constructor(public readonly stack: Stack) {
-        super('cdk:index:Stack', stack.node.id, {}, stack.options);
-        this.options = stack.options;
-        this.dependencies.push(stack.pulumiSynthesizer.stagingStack);
+    protected async initialize(props: {
+        name: string;
+        args?: AppOptions;
+        opts?: pulumi.ComponentResourceOptions;
+    }): Promise<AppResource> {
+        const cli = AwsCdkCli.fromCloudAssemblyDirectoryProducer(this);
+        this.appProps = props.args?.props;
+        this.appOptions = props.args;
+        try {
+            // TODO: support lookups https://github.com/pulumi/pulumi-cdk/issues/184
+            await cli.synth({ quiet: true, lookups: false });
+        } catch (e: any) {
+            if (typeof e.message === 'string' && e.message.includes('Context lookups have been disabled')) {
+                const message = e.message as string;
+                const messageParts = message.split('Context lookups have been disabled. ');
+                const missingParts = messageParts[1].split('Missing context keys: ');
+                error(
+                    'Context lookups have been disabled. Make sure all necessary context is already in "cdk.context.json". \n' +
+                        'Missing context keys: ' +
+                        missingParts[1],
+                    this,
+                );
+            } else {
+                error(e, this);
+            }
+        }
 
-        this.name = stack.node.id;
+        const converter = new AppConverter(this);
+        converter.convert();
 
-        const assembly = stack.app.synth();
-        this.assemblyDir = assembly.directory;
-        debug(`ASSEMBLY_DIR: ${this.assemblyDir}`);
+        return {
+            converter,
+        };
+    }
 
-        debug(JSON.stringify(debugAssembly(assembly)));
+    /**
+     * produce is called by `AwsCdkCli` as part of the `synth` operation. It will potentially
+     * be called multiple times if there is any missing context values.
```

Comment:
I thought we're exiting if it does encounter context lookups that aren't populated in `cdk.context.json`?

Reply:
Yes that is true, I'll add some more detail here saying that is how it works ideally but for the time being it will be executed only once since we are not allowing lookups.